### DISCUSSION TITLE
High precedence in sql alchemy has reversed meaning to the operator precedence in python

### DISCUSSION THREAD

**Discussion by PiotrCzapla** [sqlalchemy/sqlalchemy#8584] [Ideas]
In sqlalchemy the high value of precedence means lower precedence. While if you look at (Python's operators precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence), they state:

> The following table summarizes the operator precedence in Python, from highest precedence (most binding) to lowest precedence (least binding).

Which is a bit confusing for new comers. Would you mind sharing some insights into why it is reversed so that it is easier to remember?

I would also update the docs to state (least binding / most binding) in a few places where precedence is mentioned (like here in the [faq](https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#i-m-using-op-to-generate-a-custom-operator-and-my-parenthesis-are-not-coming-out-correctly)), along with a longer explanation to the [Operators.op.precedence](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.Operators.op.params.precedence).

Let me know what you think.

**Comment by zzzeek** [comment]
> In sqlalchemy the high value of precedence means lower precedence. While if you look at (Python's operators precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence), they state:
> 
> > The following table summarizes the operator precedence in Python, from highest precedence (most binding) to lowest precedence (least binding).
> 
> Which is a bit confusing for new comers. Would you mind sharing some insights into why it is reversed so that it is easier to remember?

hey there  -

I'm not familiar with what you are referring towards.   Our precedence table is an internal structure which you can see below:

https://github.com/sqlalchemy/sqlalchemy/blob/8c07c68c36815de6cb3aeeeb87a94824310cb551/lib/sqlalchemy/sql/operators.py#L2176-L2229

the numbering uses higher numbers for higher precedence.   That is, multiplication (*) is higher precedence than addition (+).  This matches Python's documentation at https://docs.python.org/3/reference/expressions.html and PostgreSQLs doc (where I think we most referred towards originally when making this table) at https://www.postgresql.org/docs/7.2/sql-precedence.html .

I think what might be an issue of confusion is how SQLAlchemy **uses** this precedence table, which is entirely different from what a language interpreter does.   We use it strictly to know when parenthesis need to be rendered in a SQL string, for the purposes of **explicitly binding** a particular operator that would otherwise be of lower precedence than another one we are parenthesizing against.  this is so that, given a user's Python expression with a particular explicit operator order, we can ensure that same binding order is passed along to the SQL interpreter.    That concept is certainly confusing, fortunately users of SQLAlchemy don't really need to be concerned about it in most cases.


> 
> I would also update the docs to state (least binding / most binding) in a few places where precedence is mentioned (like here in the [faq](https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#i-m-using-op-to-generate-a-custom-operator-and-my-parenthesis-are-not-coming-out-correctly)), 

Sure, higher precedence means most binding like everywhere else, the first part of that example could be worded more to indicate that we are trying to tell the compiler that our custom operator should be assumed to have higher operator precedence as interpreted by the database than the addition operator, so if we'd like `(a + b) <our op> c` to be explicitly parenthesized, we need to tell SQLAlchemy that `<our op>` on the database side has a higher precedence than `+`.

> along with a longer explanation to the [Operators.op.precedence](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.Operators.op.params.precedence).

that docstring should likely link to the FAQ at https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#why-are-the-parentheses-rules-like-this , which can be enhanced to the degree that it makes sense to newcomers.

**Comment by PiotrCzapla** [comment]
That is clear now! You nailed the source of my confusion. On introspecting it seems that
 I haven't noticed that there is already explicit parenthesis around + in the expression below:
```print((column('q1') + column('q2')).op('->')(column('p')))```
and I assumed it goes to  sql alchemy as `q1 + q2 -> p` and precedence is used to tell sqlalchemy how to parse this to an expression tree. I totally forgot that sqlalchemy is getting trees already from python parser.

Unfortunately docs to op.precedence did not help:
> precedence -  precedence to *apply* to the operator, when parenthesizing expressions. A *lower number* will cause the *expression to be parenthesized* ...

I think adding link to faq  entry you suggested will help. It would help me as well if the docstring would state more explicitly that it is just a hint how the operator will be interpreted by sql,  maybe something along this lines:

> precedence - a hint that let us understand how the sql engine will interpret this operator.  It is used when generating sql expression to minimise amount of added parenthesis see (faq). The default value is set to 0 ...

**Reply by zzzeek** [reply]
OK I've made those improvements in f35446512582defc2fcde826b8c0

**Reply by PiotrCzapla** [reply]
Thank you
